Run spec tests in parallel to reduce the execution time

stevenfontanella opened the pull request:
Reduces runtime from ~15 minutes to 1.5 minutes on my machine

Before:
<img width="830" height="155" alt="image" src="https://github.com/user-attachments/assets/afd0a52b-b311-46e6-8eb1-236295cb91de" />

After:
<img width="986" height="559" alt="image" src="https://github.com/user-attachments/assets/547eec8c-f442-486f-8cf1-49cd5eb36a64" />

Failed test example (stdout and stderr isn't ordered with the exception unfortunately, but it's easy to re-run the particular test): 
<img width="1707" height="929" alt="image" src="https://github.com/user-attachments/assets/dc1443d7-fe6c-4d23-ac63-44da4fd92c90" />

* Run tests through a thread pool with `os.cpu_count() * 2` threads
  * `os.cpu_count() * 4` shows no benefit. `os.cpu_count()` or `os.cpu_count() // 2` also show no regression in runtime, but might be worse for machines with less cores. There are currently 315 spec tests total for reference.
* Prefixes round-trip file name tests with their test name to avoid clobbering the a.wasm / ab.wast files during tests
* Add stdout and stderr params to functions that print so that lines can be captured by each thread and not interleaved
  * Note that we pass stdout as the stderr param in practice so that they are interleaved, otherwise all stdout lines and stderr lines will be outputted together in each test.

kripken commented on check.py:
```diff
@@ -175,73 +181,124 @@ def run_wasm_reduce_tests():
+        for i, (module, asserts) in enumerate(support.split_wast(str(wast))):
+            if not module:
+                # Skip any initial assertions that don't have a module
+                return
```
This looks wrong. Before we continued inside this loop - which is unchanged in this PR - but now we return from the entire function. That is, continues from the outer loop made sense to turn into returns, but this is an inner loop?

stevenfontanella commented on check.py:
```diff
@@ -175,73 +181,124 @@ def run_wasm_reduce_tests():
+    printing_thread = threading.Thread(target=printer)
+    printing_thread.start()
+
+    worker_count = os.cpu_count() * 2
```
It doesn't in my case. I put some details on this in the description; even cpu_count // 2 shows the same performance for me. This is a maximum, so probably the threads are progressing fast enough that we don't end up needing to spawn all of them.

I picked this number because I suspect that machines with a smaller cpu count than mine (128) would want more. Some of the spec tests take about as long as the entire test suite (return_call_ref.wast takes 1.5 minutes by itself), so it's important that those tests get scheduled ASAP. And I guess this is also why the thread count doesn't matter much in my case; we're bottlenecked by return_call_ref on its own.

stevenfontanella replied:
SGTM, I don't have a strong reason to double it

stevenfontanella commented on check.py:
```diff
@@ -175,73 +181,124 @@ def run_wasm_reduce_tests():
+            shared.fail(actual, expected)
+
+
+def run_one_spec_test(wast: Path, stdout=None, stderr=None):
```
I think Path is generally better since it conveys what this parameter really is and ensures that the caller passes a real path instead of a random string. I also use Path here so we can do `wast.parts` ([explained here](https://github.com/WebAssembly/binaryen/pull/8088#discussion_r2590169934)) which os.path doesn't seem to have (we could do path.split("/") but this is wrong on Windows and we might as well use the common library)

stevenfontanella commented:
Re:

> The alpine builder has been running for over three hours now, so I think there's a problem here. Looking at the log, it looks much more verbose than before because the stdout from the wasm-shell commands is being printed. Maybe fixing that will make the builder go faster?

I changed `support.run_command` to not print the process's stdout and fixed the Alpine build (it's using a lower Python version and didn't have `Queue.shutdown()`)

stevenfontanella commented:
I think the CI problem is that [this member](https://github.com/WebAssembly/binaryen/blob/main/src/support/small_vector.h#L52) isn't initialized in the default constructor of SmallVector since std::array is an aggregate type. I'm not sure why we see the breakage here and not in main or in other architectures. I'll try to re-run for now and send a PR to fix separately.

[CI link](https://github.com/WebAssembly/binaryen/actions/runs/19947946325/job/57201682487?pr=8088)

kripken commented:
About the `std::array` initialization error, we do disable that warning a few lines above for some archs, and maybe we just need to disable it in more. gcc does seem to have many such false positives...

kripken commented on src/support/small_vector.h:
```diff
@@ -49,7 +49,7 @@ namespace wasm {
   // fixed-space storage
   size_t usedFixed = 0;
-  std::array<T, N> fixed;
+  std::array<T, N> fixed{};
```
Let's move this to a separate PR. Can we remove the pragmas that disable the warning, above, after it?
